[exporter/fileexporter] add periodic flushing

peterzandbergen opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
Added flushing to the fileExporter to ease live debugging.
Added flush_interval parameter to the Config struct. 
The value of flush_interval should be 1 or higher. 
If not flushing is disabled.
Flushing occures every flush_interval seconds.

**Link to tracking Issue:**
Resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18251

**Testing:** <Describe what testing was performed and which tests were added.>
Tests were added to config_test.go, file_exporter_test.go. These passed.
Test were also performed by running the otelcontrib program. 

**Documentation:** <Describe the documentation added.>
The README.md file has been updated with the new flush_interval setting.

peterzandbergen committed:
Add flushing to the fileexporter

peterzandbergen committed:
Add flushing to fileexporter

Add flush_interval in seconds to the fileexporter configuration
and use this parameter to periodically flush the buffer to the file.

Flushing eases debugging because the output is shown at regularly
intervals.

Not setting the flush_interval or setting it to a value less than 1
results in the old behavior.

djaglowski commented:
@peterzandbergen, thanks for contributing this. Please sign the CLA.

We can use https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18251 as the tracking issue since this need was noted there as well.

peterzandbergen committed:
Add tests and changelog yaml

peterzandbergen commented:
CLA is signed

djaglowski commented on exporter/fileexporter/file_exporter.go:
```diff
@@ -112,15 +116,57 @@ func exportMessageAsBuffer(e *fileExporter, buf []byte) error {
+
+// Start starts the flush timer.
 func (e *fileExporter) Start(context.Context, component.Host) error {
+	startFlusher(e)
```
I would prefer that we determine here whether or not to call `startFlusher` and that we do so by comparing to the user specified value. I think this keeps the business logic clearer.

```go
if e.flushInterval > 0 {
  e.startFlusher()
}
```

peterzandbergen replied:
Changed

djaglowski commented on exporter/fileexporter/file_exporter.go:
```diff
@@ -112,15 +116,57 @@ func exportMessageAsBuffer(e *fileExporter, buf []byte) error {
 	return nil
 }
 
+func startFlusher(e *fileExporter) {
```
Rather than pass the exporter into the function, let's make this a receiver method.
```suggestion
func (e *fileExporter) startFlusher() {
```

peterzandbergen replied:
Understood

djaglowski commented on exporter/fileexporter/file_exporter.go:
```diff
@@ -112,15 +116,57 @@ func exportMessageAsBuffer(e *fileExporter, buf []byte) error {
+	}()
+}
+
+func stopFlusher(e *fileExporter) {
```
```suggestion
func (e *fileExporter) stopFlusher() {
```

peterzandbergen replied:
Changed

djaglowski commented on exporter/fileexporter/file_exporter.go:
```diff
@@ -112,15 +116,57 @@ func exportMessageAsBuffer(e *fileExporter, buf []byte) error {
 // Shutdown stops the exporter and is invoked during shutdown.
 func (e *fileExporter) Shutdown(context.Context) error {
+	// Stop the flush timer.
+	stopFlusher(e)
```
```suggestion
	if e.flushTicker != nil {
		stopFlusher(e)
	}
```

peterzandbergen replied:
Changed

djaglowski commented on exporter/fileexporter/testdata/config.yaml:
```diff
@@ -39,3 +39,7 @@ file/format_error:
+
+file/flush_interval_5:
+  path: ./flushed
+  flush_interval: 5
```
```suggestion
  flush_interval: 5

```

peterzandbergen replied:
I have added test cases for different units for flush_interval.

djaglowski commented on exporter/fileexporter/config.go:
```diff
@@ -44,6 +44,9 @@ type Config struct {
 	Compression string `mapstructure:"compression"`
+
+	// FlushInterval number of seconds between flushes.
+	FlushInterval int `mapstructure:"flush_interval"`
```
Typically we will use a `time.Duration` for measures of time. This allows users to pick more aggressive intervals if they want, or easily use very long ones. e.g. `100ns`, `2h`

peterzandbergen replied:
I  have changed it to Duration.

atingchen commented on exporter/fileexporter/file_exporter.go:
```diff
@@ -112,15 +116,57 @@ func exportMessageAsBuffer(e *fileExporter, buf []byte) error {
+	if e.flushInterval < 1 {
+		// Interval should be 1 or larger.
+		return
+	}
```
When `Start` calls `startFlusher`, I don't think there will be a situation where e.flushTicker != nil. As djaglowski said, we should determine whether to call the `startFlusher`  when the `Start` is called. So, can these two if statements be deleted?

peterzandbergen replied:
You have a point. I have changed it.

peterzandbergen committed:
Add new meaning of flush_interval

peterzandbergen committed:
Process reviewers comments and add stop procedure

- Change flush_interval to time.Duration
- Change startFlusher to a receiver method
- Moved stopFlusher logic to Stop method
- Move test on flush_interval to Start method
- Remove superfluous tests
- Add stopFlusher channel to ensure that the flushing goroutine
  stops correctly

peterzandbergen commented:
Hi Daniel and Iook (I hope this is first name),
Thanks for your comments and suggestions for improvement. I have incorporate them all.
I have also improved the way the goroutine stops.

Changes
- Change flush_interval to time.Duration
- Change startFlusher to a receiver method
- Moved stopFlusher logic to Stop method
- Move test on flush_interval to Start method
- Remove superfluous tests
- Add stopFlusher channel to ensure that the flushing goroutine
  stops correctly

peterzandbergen committed:
Fix error on new line in config.yaml

peterzandbergen commented:
Hi, can you let me know if I need to perform more changes? Cheers

djaglowski commented:
Thanks for making this contribution @peterzandbergen. The code looks good to me and I've verified manually that it works as expected. 

There are two things I think we need to consider yet. 
1. Should we enable this functionality by default? I think probably yes, as this component is commonly used for manual testing and the lack of automatic flushing often causes confusion. I don't see any notable downsides to having an unaggressive flush interval. I think `1s` would be reasonable. WDYT? Perhaps @atingchen, do you have an opinion on this?
2.  It's not clear to me that any of the tests are actually validating this functionality. I think we should have a test that ensures the flush occurs. Apologies if I missed this but I couldn't find it.

peterzandbergen commented:
Hi @djaglowski,
thanks for your comment. 

ad 1) I agree that a default flush interval makes sense. When I started using the fileexporter I thought I was doing something wrong when I saw no output. A default of 1 second sounds reasonable.

ad 2) I will think of a test to validate if flushing works.

atingchen commented:
> 1. Should we enable this functionality by default? I think probably yes, as this component is commonly used for manual testing and the lack of automatic flushing often causes confusion. I don't see any notable downsides to having an unaggressive flush interval. I think `1s` would be reasonable. WDYT? Perhaps @atingchen, do you have an opinion on this?

I think this sounds like a good idea.

peterzandbergen committed:
Add flush and buffer to rotation mode

Also
- set flush_interval to 1s if not set
- use buffered writer for rotation logger
- add flush interval test

peterzandbergen committed:
Change nopWriteCloser to writeNopCloser

peterzandbergen committed:
Add default flush_interval value to README

peterzandbergen committed:
Fix failing tests

- Set expected default value in config_test
- Add unwrapping to the rotation settings test
- Reuse the NopWriterCloser instead of adding my version

peterzandbergen committed:
Fix race condition in fileexporter test

peterzandbergen committed:
Fix lint errors in fileexporter_test

djaglowski commented:
@peterzandbergen, I'll be there and look forward to connecting.

peterzandbergen committed:
Fix config_test formatting and fix race condition

- lint complained about bad formatting
- add mutex to startFlushing to avoid race condition

atingchen commented:
@peterzandbergen Thank you very much for your invitation. Unfortunately, I won't be able to make it to the event.Hope you have a happy memory.

peterzandbergen commented:
@atingchen Hi Luke, it was great working together. Let's do it again sometime.

peterzandbergen committed:
Fix race conditions

- add mutex lock/unlock to shutdown
- add safeFileExporterWrite with mutex lock/unlock

peterzandbergen commented:
I am adding commits to fix race errors in testing.

peterzandbergen commented:
@djaglowski Hi, I see that you have approved it. I is now waiting for another reviewer. Is there something I need to provide for it to be approved?
I am asking because I am not familiar with the whole process.
Cheers

BTW, are you attending KubeCon as a speaker or does your company have a booth?

djaglowski commented on exporter/fileexporter/file_exporter.go:
```diff
@@ -134,3 +182,4 @@ func buildExportFunc(cfg *Config) func(e *fileExporter, buf []byte) error {
 	}
 	return exportMessageAsLine
 }
+
```
```suggestion
```

djaglowski committed:
Fix lint

djaglowski commented:
@peterzandbergen, thanks for resolving the race condition. There was one more lint issue but I believe I've resolved it. I'll reach out on CNCF Slack re KubeCon.

@atingchen, do you want to give this a final review?

peterzandbergen commented:
@djaglowski @atingchen Dan and Luke, 
I want to thank you for your cooperation and patience in finishing this PR. 
I was not familiar with the way the project was set up, especially I needed to learn about the different options for make for lint and test and to localize this to the exporter/fileexporter directory. 
All in all it was a good experience and it encourages me to contribute more. 
Again, thanks.

peterzandbergen commented:
One last thing, I hope. There is no check on a negative duration for flush_interval, it will result in panic when the ticker is created. I suggest to return an error in this case and document it.

peterzandbergen committed:
Add error for negative flush_interval

Avoid panic

atingchen commented:
> @peterzandbergen, thanks for resolving the race condition. There was one more lint issue but I believe I've resolved it. I'll reach out on CNCF Slack re KubeCon.
> 
> @atingchen, do you want to give this a final review?

It looks good to me, but I'm not authorized to approved it.

EdmundMartin commented:
Is anyone actively looking at this to provide the second review? Found the fact that the file exporter doesn't periodically flush to disk quite surprising and this PR would be a good fit for my use case. Would love to see this get merged in!

Is there anything that can be done to get this change over the line?

djaglowski commented:
Thanks for the contribution @peterzandbergen!
